### PR TITLE
edge/devicetwin: add unit tests for membership manager

### LINKED ISSUE #6956: tests: add unit tests for edge/pkg/devicetwin/dtmanager/membership.go
**What would you like to be added/modified:**

`edge/pkg/devicetwin/dtmanager/membership.go` (~428 LOC) has no `_test.go` counterpart. It handles device membership events (connect, disconnect, device updates) on the edge.

Suggested coverage:
- `dealMembershipGet` — returns correct membership for a known device group, empty result for unknown
- `dealMembershipUpdated` — processes add/remove device events correctly
- `dealMembershipDetail` — correct detail response for connected devices
- Error paths where `DeviceServiceFactory` mock returns failure on `AddDevices` / `DeleteDevices`

**Why is this needed:**

`DeviceServiceFactory` is already an injectable function variable (documented in `device.go:26`), making mock injection easy. Membership events are on the critical edge path — a regression here silently desynchronizes the device list between cloud and edge.

Tracked as part of #6927.

/kind test

### PR DESCRIPTION
**Pull request by Priyanshubhartistm**
**What type of PR is this?**

/kind test
/kind bug

**What this PR does / why we need it**:

`edge/pkg/devicetwin/dtmanager/membership.go` (~428 LOC) had no `_test.go` counterpart. Membership events are on the critical edge path — a regression here silently desynchronises the device list between cloud and edge.

This PR adds comprehensive unit tests for all public functions in `membership.go` using the injectable `MembershipServiceFactory` variable.

While writing the tests, two production bugs were discovered and fixed:

**MockDeviceService bug (`mock_device.go`):** `AddDeviceTrans` and `DeleteDeviceTrans` hard-coded `return nil` instead of delegating to their overridable `AddDeviceTransFunc` / `DeleteDeviceTransFunc` fields. This made error injection impossible for negative-path tests.

**Unlock-without-lock panic (`membership.go`, `delta=false` production path):** In `addDevice()`, when `AddDeviceTrans` fails and `delta == false`, the device mutex was never locked, but `context.Unlock(device.ID)` was called unconditionally — causing a latent panic. Fixed by guarding the `Unlock` call with the same `delta` condition used for the preceding `Lock`. Additionally, `DeviceMutex.Delete(device.ID)` was called before `context.Unlock(device.ID)` in both `addDevice()` and `removeDevice()`. Since `Unlock()` looks up the mutex via `DeviceMutex.Load()`, deleting it first made `Unlock()` a silent no-op. Fixed by calling `Unlock()` before `Delete()` for the `delta=false` production path. The `delta=true` concurrent-access lifecycle (where another goroutine may already hold a reference to the mutex before `DeviceMutex.Delete` is called) is not fully addressed in this PR and is deferred to a follow-up.

**Note on `delta=true` callers:** All current production call sites invoke `addDevice` / `removeDevice` with `delta=false` (via `dealMembershipUpdate` and `dealMembershipDetail`). The `delta=true` paths are tested for correctness of the lock/unlock bracket but are not currently exercised by any production caller; this is documented for reviewers.

**Test coverage added**:

- `dealMembershipGet` — valid payload, invalid message type, invalid content, known device returns correct membership list
- `dealMembershipUpdate` — add device success, remove device success, `AddDeviceTrans` failure (device cleaned from context, no panic), `DeleteDeviceTrans` failure
- `dealMembershipDetail` — valid payload, stale-device removal path
- `SyncDeviceFromSqlite` — `QueryDevice` error, device not found, `QueryDeviceAttr` error, `QueryDeviceTwin` error, success path
- `MemWorker.Start` — valid action dispatch, unknown action dispatch, heartbeat channel closed
- `initMemActionCallBack` — all three callbacks registered
- `addDevice` with `delta=true` error path — verifies `Unlock` is called before `DeviceMutex.Delete`
- `removeDevice` with `delta=true` — verifies Lock/Unlock bracket and map entry cleanup

**Which issue(s) this PR fixes**:

Fixes #6956

**Special notes for your reviewer**:

The two bug fixes are minimal-scope changes. The `make verify` CRD check failure is pre-existing and unrelated to this PR (tracked separately).

All `./edge/pkg/devicetwin/...` tests pass with **80.6% statement coverage** on the `dtmanager` package.

**Does this PR introduce a user-facing change?**:

```release-note
fix: membership manager no longer panics on AddDeviceTrans failure when delta=false; DeviceMutex entry is now cleaned up after Unlock to prevent silent no-op unlock
```

### CODE REVIEW

File: edge/pkg/devicetwin/dtmanager/membership.go [hunk 1]
```diff
@@ -237,7 +237,9 @@ func addDevice(context *dtcontext.DTContext, toAdd []dttype.Device, baseMessage
 		if err != nil {
 			klog.Errorf("Add device %s failed due to some error ,err: %#v", device.ID, err)
 			context.DeviceList.Delete(device.ID)
-			context.Unlock(device.ID)
+			if delta {
+				context.Unlock(device.ID)
+			}
```
[the remark this answers is not in the document]
Comment (Priyanshubhartistm) [reply edge/pkg/devicetwin/dtmanager/membership.go:242]: done

File: edge/pkg/devicetwin/dtmanager/membership_test.go [hunk 2]
```diff
@@ -36,261 +36,328 @@ var (
...
-	}
-	dArray = append(dArray, d)
+	dArray := []dttype.Device{{ID: "123"}}
 	value := getRemoveList(dtc, dArray)
 	for i := range value {
 		if value[i].ID != "DeviceB" {
-			t.Errorf("expected %v, but got %v", "DeviceB", value[i].ID)
+			t.Errorf("expected DeviceB, got %v", value[i].ID)
 		}
 	}
```
[the remark this answers is not in the document]
Comment (Priyanshubhartistm) [reply edge/pkg/devicetwin/dtmanager/membership_test.go:71]: Address in this commit https://github.com/kubeedge/kubeedge/pull/6990/commits/545811d3df952f7c8ffa7bce2a86095ff874a834

**Comment by Priyanshubhartistm** [comment]
Hey @DoisLONG , could you please take a look at this PR? whenever you have sometime, I would really appreciate your review and any feedback or suggestions you may have.

Thank you

**Comment by DoisLONG** [comment]
Thanks for adding the membership tests. The coverage direction is useful, and I also see the existing inline comments around cleanup and stronger assertions have mostly been addressed.

One issue I think still needs to be fixed before LGTM: in the `addDevice()` error path, `context.DeviceMutex.Delete(device.ID)` is called before `context.Unlock(device.ID)`.

Since `DTContext.Unlock()` first looks up the mutex from `DeviceMutex`, deleting the entry before calling `Unlock()` means the unlock path will not actually find and unlock the mutex when `delta == true`.

Could you change the order to unlock first, then delete the mutex entry? For example:

```go
context.DeviceList.Delete(device.ID)
if delta {
    context.Unlock(device.ID)
}
context.DeviceMutex.Delete(device.ID)
```

Also, since this PR includes a production bug fix in addition to tests, it may be better to add `/kind bug` as well, or otherwise make that scope clear in the PR metadata.

**Comment by Priyanshubhartistm** [comment]
Hi @DoisLONG, thank you for the review!

I've fixed the ordering in both addDevice and removeDevice so that Unlock is called before DeviceMutex.Delete. Since DTContext.Unlock() internally looks up the mutex via DeviceMutex.Load(), deleting the entry first would cause Unlock to silently no-op, leaving the sync.Mutex permanently locked.

I've also added /kind bug to the PR metadata to reflect that this includes a production bug fix alongside the unit tests.

Please let me know if anything else needs to be addressed!

**Comment by DoisLONG** [comment]
Thanks for the update.

The code changes look good to me now. The previous concern around the `Unlock()` / `DeviceMutex.Delete()` ordering has been addressed, and the PR metadata now also reflects that this includes a bug fix.

One remaining item before LGTM: the current failing check appears to be `codecov/patch`, with the diff coverage below the target. Could you please either improve the test coverage for the changed production lines, or clarify why this coverage drop is acceptable?

Also, please fix the release-note block formatting in the PR description. It should be:

```release-note
NONE
```

Once the coverage/check issue and release-note formatting are resolved, this looks good to me.

**Review by DoisLONG — COMMENTED** [review]
Thanks for the updates.

The main code concerns look addressed to me now. The `addDevice()` and `removeDevice()` cleanup paths now unlock before deleting from `DeviceMutex`, which avoids the previous no-op unlock / leaked lock problem. The added assertions around `DeviceList` and `DeviceMutex` cleanup also make the regression coverage stronger.

The mock device service change also makes sense, since `AddDeviceTrans` and `DeleteDeviceTrans` now delegate to the injectable mock functions instead of always returning nil.

One remaining thing to check: the current Main CI workflow is still marked as failed, although the edge unit/integration and verify jobs I could see are passing. Could you take a look at the failing workflow state, or rerun it if this is unrelated/flaky?

Also, since this PR now includes actual bug fixes in addition to tests, the release note probably should not stay as `NONE`. A short note about fixing the membership manager mutex cleanup / unlock ordering would be more accurate.

**Comment by Priyanshubhartistm** [comment]
Thanks @DoisLONG , for the review! I've made the requested updates, including the release note, and checked the CI status. 
Please let me know if you have any suggestions

**Review by DoisLONG — COMMENTED** [review]
Thanks for the update. I took another pass specifically through the lock lifecycle, and I do not think the current ordering change or the new tests fully prove the claimed fix yet.

`DTContext.Lock()` loads the mutex from `DeviceMutex` before acquiring the global `RLock`. Therefore, moving `DeviceMutex.Delete()` after `Unlock()` still leaves a race window: another goroutine may already have loaded the same mutex and be waiting on it. After the remover unlocks and deletes the map entry, that goroutine can acquire the old mutex, but its later `Unlock(deviceID)` will no longer find it in `DeviceMutex`, leaking both the device mutex and the global read lock.

The two new delta=true tests do not catch this. They only verify that the map entries were deleted, which would also pass with the previous Delete-before-Unlock implementation. The race detector will not detect a leaked lock or deadlock automatically.

Could you please either:

1. explain why the delta=true paths cannot run concurrently with other device handlers, or
2. adjust the mutex lifecycle so an in-flight waiter cannot lose its map entry?

Please also make the regression test verify that the original device mutex is actually unlocked and that the global write lock can still be acquired after cleanup.

One related point: the current production call sites pass `delta=false`, so please clarify the real caller or trigger for the delta=true issue in the PR description.

**Comment by Priyanshubhartistm** [comment]
hii @DoisLONG ,  could you please take another look at this PR? whenever you have sometime, I would really appreciate your review and any feedback or suggestions you may have.

Thank you

**Review by DoisLONG — COMMENTED** [review]
Thanks for the follow-up. I took another look at the latest version.

The membership test coverage and the mock service update look useful to me. The `addDevice()` error path for the current `delta=false` production callers also looks better now, since it no longer tries to unlock a device mutex that was not locked in that path.

One thing I still want to clarify: I do not think the `delta=true` mutex cleanup should be described as fully fixed yet. `DTContext.Lock()` loads the device mutex from `DeviceMutex` before acquiring the global `RLock()`, and `DTContext.Unlock()` loads it again before calling `RUnlock()`. So if another goroutine has already loaded the old mutex and is waiting on it, deleting the map entry after the current unlock can still make the later `Unlock(deviceID)` fail to find the mutex.

Since the PR description now says the current production callers use `delta=false`, I think the cleanest path is to narrow the description a bit: keep this PR focused on the real production fix, the mock-device-service fix, and the membership test coverage. For the `delta=true` lock lifecycle, either avoid claiming it is fully fixed here, or handle it in a separate follow-up with a stronger concurrency-safe design/test.

**Comment by Priyanshubhartistm** [comment]
thanku for your review @DoisLONG , could you please take another look at this PR when you have time
If you have any suggestions, please let me know.

**Review by DoisLONG — COMMENTED** [review]
Thanks for the update.

One small cleanup: please adjust the comments in the `delta=true` tests. They currently say things like “the race detector will catch it” or “the race detector verifies the ordering is correct,” but the race detector does not prove leaked-lock or in-flight-waiter safety. Since the PR now explicitly defers the full `delta=true` lifecycle issue, the test comments should also avoid claiming that this path is fully proven. It is enough to say these tests cover the current cleanup behavior, while the stronger concurrent lifecycle is left for follow-up.

After that cleanup, this looks reasonable to me.

**Comment by Priyanshubhartistm** [comment]
thanks @DoisLONG  good catch, please let me know if you have any other suggestion

**Review by DoisLONG — COMMENTED** [review]
Thanks for the updates and for iterating on this PR.

I checked the latest version again. The main concerns from the previous rounds look addressed to me:

- `addDevice()` no longer calls `Unlock()` on the `delta=false` failure path.
- `DeviceMutex` cleanup is covered when `AddDeviceTrans` fails.
- `removeDevice()` now unlocks before deleting the mutex entry.
- The membership manager test coverage is much more complete now, including add/remove paths, sqlite sync, membership get/detail/update, worker start, and the `delta=true` cleanup paths.

The note about the stronger concurrent lifecycle for `delta=true` being deferred to a follow-up is acceptable to me, since the current production callers use the `delta=false` path.

/lgtm

**Comment by Priyanshubhartistm** [comment]
Thank you for the review, @DoisLONG , @WillardHu
If everything looks good now, could you please /approve this PR when you have a chance? I'd really appreciate it.
